homeassistant/components/balboa/climate.py

```diff
@@ -0,0 +1,207 @@
+        if self._client.get_heatmode() == self._client.HEATMODE_RNR:
+            return [*CLIMATE_SUPPORTED_MODES, HVAC_MODE_AUTO]
+        return CLIMATE_SUPPORTED_MODES
+
+    @property
+    def hvac_mode(self) -> str:
+        """Return the current HVAC mode."""
+        mode = self._client.get_heatmode()
+        return self.balboa_to_ha_heatmode_map[mode]
+
+    @property
+    def hvac_action(self) -> str:
+        """Return the current operation mode."""
+        state = self._client.get_heatstate()
+        if state >= self._client.ON:
+            return CURRENT_HVAC_HEAT
+        return CURRENT_HVAC_IDLE
+
+    @property
+    def fan_mode(self) -> str:
+        """Return the current fan mode."""
+        fanmode = self._client.get_blower()
+        if fanmode in self.balboa_to_ha_blower_map:
+            return self.balboa_to_ha_blower_map[fanmode]
+        return FAN_OFF
+
+    @property
+    def precision(self) -> float:
+        """Return the precision of the system.
+
+        Balboa spas return data in C or F depending on how the display is set,
+        because ultimately, we are just reading the display.
+        In C, we have half-degree accuracy, in F, whole degree.
+        """
+        if self.hass.config.units.temperature_unit == TEMP_CELSIUS:
+            return PRECISION_HALVES
+        return PRECISION_WHOLE
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement, as defined by the API."""
+        tscale = self._client.get_tempscale()
+        if tscale == self._client.TSCALE_C:
+            return TEMP_CELSIUS
+        return TEMP_FAHRENHEIT
+
+    @property
+    def current_temperature(self):
+        """Return the current temperature."""
+        return self._client.get_curtemp()
+
+    @property
+    def target_temperature(self):
+        """Return the target temperature we try to reach."""
+        return self._client.get_settemp()
+
+    @property
+    def min_temp(self) -> int:
+        """Return the minimum temperature supported by the spa."""
+        trange = self._client.get_temprange()
+        scale = self._client.get_tempscale()
+        return self._client.tmin[trange][scale]
+
+    @property
+    def max_temp(self) -> int:
+        """Return the minimum temperature supported by the spa."""
+        trange = self._client.get_temprange()
+        scale = self._client.get_tempscale()
+        return self._client.tmax[trange][scale]
+
+    @property
+    def preset_modes(self):
+        """Return the valid preset modes."""
+        modes = [
+            mode
+            for mode in self._client.get_heatmode_stringlist()
+            # only return "Ready in Rest" as an option if the spa is currently in that
+            # mode since it is a status rather than an available mode to be selected
+            if self._client.get_heatmode() == self._client.HEATMODE_RNR
+            or mode != self._client.get_heatmode_stringlist()[self._client.HEATMODE_RNR]
```

Comment:
The list of preset modes isn't allowed to change at run time (it may break HomeKit, Alexa, Google Assistant or any other integration that consumes these).  It should be set once when the entity is created.

Reply:
Ideally this are set in `__init__` as `self._attr_preset_modes`

Reply:
OK, so, this one is a problem.  Thats simply not how the spa itself physically works.  You can never put the spa into that mode, and if you show that option, and the user selects it when the spa is not in that mode, ha and the spa diverge from reality.  If you decide to just never show that mode, you can't tell when the spa is in the mode.

In HACS, my original setup was just just have this preset like you said as all three.  Every user complained that it made the integration a brick for them, because when the spa moved through modes it would be wrong in HA, and then you would lose control of that feature of the spa (which is kind of important if you actually want to automate the spa in any reasonable way).  The way the device physically handles this mode setting internally, dictates that it operates like this.